jina/drivers/multimodal.py

```diff
@@ -0,0 +1,112 @@
+__copyright__ = "Copyright (c) 2020 Jina AI Limited. All rights reserved."
+__license__ = "Apache-2.0"
+
+from collections import defaultdict
+from typing import Iterable, Tuple, Dict
+
+import numpy as np
+
+from .encode import BaseEncodeDriver
+from .helper import pb2array, array2pb
+from ..proto import jina_pb2
+
+
+def _extract_doc_content(doc: 'jina_pb2.Document'):
+    """Returns the content of the document with the following priority:
+    If the document has an embedding, return it, otherwise return its content.
+    """
+    if doc.embedding.buffer:
+        return pb2array(doc.embedding)
+    else:
+        return doc.text or doc.buffer or (doc.blob and pb2array(doc.blob))
+
+
+def _extract_modalities_from_document(doc: 'jina_pb2.Document'):
+    """Returns a dictionary of document content (embedding, text, blob or buffer) with `modality` as its key
+    """
+    doc_content = {}
+    for chunk in doc.chunks:
+        modality = chunk.modality
+        if modality in doc_content.keys():
+            return None
+        else:
+            doc_content[modality] = _extract_doc_content(chunk)
+    return doc_content
+
+
+class MultimodalDriver(BaseEncodeDriver):
+    """Extract multimodal embeddings from different modalities.
+    Input-Output ::
+        Input:
+        document:
+                |- chunk: {modality: mode1}
+                |
+                |- chunk: {modality: mode2}
+        Output:
+        document: (embedding: multimodal encoding)
+                |- chunk: {modality: mode1}
+                |
+                |- chunk: {modality: mode2}
+
+    .. note::
+        - It traverses on the ``documents`` for which we want to apply the ``multimodal`` embedding. This way
+        we can use the `batching` capabilities for the `executor`.
+        - It assumes that every ``chunk`` of a ``document`` belongs to a different modality.
+    """
+
+    def __init__(self,
+                 traversal_paths: Tuple[str] = ('r',), *args, **kwargs):
+        super().__init__(traversal_paths=traversal_paths, *args, **kwargs)
+
+    @property
+    def position_by_modality(self):
+        if not getattr(self._exec, 'position_by_modality', None):
+            raise RuntimeError('Could not know which position of the ndarray to load to each modality')
+        return self._exec.position_by_modality
+
+    def _get_executor_input_arguments(self, content_by_modality: Dict[str, 'np.ndarray']):
+        """
+            From a dictionary ``content_by_modality`` it returns the arguments in the proper order so that they can be
+            passed to the executor.
+        """
+        return [content_by_modality[modality] for modality in self.position_by_modality.keys()]
```

Comment:
The position information is not used. 
```suggestion
        return [v for k, v in sorted(self.position_by_modality.items(), key=lambda item: item[1])]
```

Reply:
After setting `self.position_by_modality = {'visual1': 1, 'visual2': 0, 'textual': 2}`, the following test will fail.
```
def test_multimodal_driver_with_shuffled_order(simple_multimodal_driver, mock_multimodal_encoder, doc_with_multimodal_chunks):
    simple_multimodal_driver.attach(executor=mock_multimodal_encoder, pea=None)
    simple_multimodal_driver._apply_all([doc_with_multimodal_chunks])
    doc = doc_with_multimodal_chunks
    assert len(doc.chunks) == 3
    visual1 = doc.chunks[1]
    visual2 = doc.chunks[0]
    textual = doc.chunks[2]
    control = np.concatenate([pb2array(visual2.embedding), pb2array(visual1.embedding), pb2array(textual.embedding)])
    test = pb2array(doc.embedding)
    np.testing.assert_array_equal(control, test)
```

Reply:
will check!

Reply:
Done Nan, added test, it was a changed we made trying to optimize, before was good